crates/simulate/src/state/mod.rs

```diff
@@ -1,249 +1,367 @@
-                "Error while simulating evm tx: {:?}",
-                e
-            ))),
-        }
+    pub fn execute_evm_tx_and_decode<D: Detokenize>(
+        &mut self,
+        caller: Address,
+        call: ContractCall<StateMiddleware, D>,
+    ) -> StateResult<D> {
+        let tx = build_revm_tx(caller, call.tx)?;
+        let output_bytes = self.execute_evm_tx(tx)?.parse_success();
+        decode_output(&call.function, output_bytes).map_err(StateError::FunctionOutputDecodeError)
     }
 
-    pub fn execute_world_update(&mut self, update: U) -> Option<U> {
-        self.world.execute(update)
+    pub fn execute_evm_tx_state_update(
+        &mut self,
+        evm_tx_state_update: EvmStateUpdate,
+    ) -> StateResult<()> {
+        match evm_tx_state_update {
+            EvmStateUpdate::Execute(caller, tx, function, result_sender) => {
+                let tx = build_revm_tx(caller, tx)?;
+                let output = self.execute_evm_tx(tx)?;
+                result_sender
+                    .send(EvmStateUpdateOutput::Execute(output, function))
+                    .map_err(|_| StateError::ChannelSendError)?;
+            }
+            EvmStateUpdate::ExecuteRaw(_, tx, result_sender) => {
+                let output = self.execute_evm_tx(tx)?;
+                result_sender
+                    .send(EvmStateUpdateOutput::ExecuteRaw(output))
+                    .map_err(|_| StateError::ChannelSendError)?;
+            }
+        }
+        Ok(())
     }
 
-    pub fn insert_into_update_results(
+    pub fn execute_world_state_update(
         &mut self,
-        tag: Cow<'static, str>,
-        address: Address,
-        result: SimUpdateResult<U>,
-    ) {
-        if let Some(agent_update_results) = self.update_results.get_mut(&address) {
-            agent_update_results.insert(tag, result);
-        } else {
-            self.update_results
-                .insert(address, HashMap::from([(tag, result)]));
+        world_state_update: WorldStateUpdate<U>,
+    ) -> StateResult<()> {
+        let result = self.world.execute(world_state_update.update);
+        if let Some(result) = result {
+            world_state_update
+                .result_sender
+                .send(result)
+                .map_err(|_| StateError::ChannelSendError)?;
         }
+        Ok(())
     }
 
-    pub fn clear_all_results(&mut self) {
-        for results in self.update_results.drain() {
-            let address = format!("{:2X}", results.0);
-            for result in results.1.iter() {
-                log::debug!("{}({}): {:?}", result.0, address, result.1);
+    pub fn execute_call_raw<D: Detokenize>(&mut self, tx: TxEnv) -> StateResult<EvmBytes> {
+        self.evm.env.tx = tx;
+
+        let output = self
+            .evm
+            .transact_commit()
+            .map_err(StateError::EvmExecutionError)?;
+        Ok(unpack_evm_tx_output(output)?.parse_success())
+    }
+
+    pub fn deploy_evm_contract<T: Tokenize>(
```

Comment:
This replaces the previous functionality from `SimContract` for deploying contracts. Much easier to call it directly during agent activation steps, rather than creating a separate `SimContract` object and deploying from there.